src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,340 @@
+using System;
+using System.Buffers;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Orleans.Runtime.Configuration.Options;
+
+namespace Orleans.Runtime.Placement;
+
+// details: https://www.ledjonbehluli.com/posts/orleans_resource_placement_kalman/
+internal sealed class ResourceOptimizedPlacementDirector : IPlacementDirector, ISiloStatisticsChangeListener
+{
+    private readonly record struct ResourceStatistics(float? CpuUsage, float? AvailableMemory, long? MemoryUsage, long? TotalPhysicalMemory, bool IsOverloaded);
+
+    /// <summary>
+    /// 1 / (1024 * 1024)
+    /// </summary>
+    private const float PhysicalMemoryScalingFactor = 0.00000095367431640625f;
+    private const int OneKiloByte = 1024;
+
+    private readonly ResourceOptimizedPlacementOptions _options;
+    private readonly ConcurrentDictionary<SiloAddress, ResourceStatistics> _siloStatistics = [];
+
+    private readonly DualModeKalmanFilter<float> _cpuUsageFilter = new();
```

Comment:
Surely, these should be per-silo in the cluster, not one for the whole cluster.

Reply:
Since I assume this is the case, I've updated the PR to use per-silo filters.

Reply:
awesome, thanks